### ISSUE TITLE
taxonomy diff validated taxonomy produces: Generating dataset failed with the following error: Error: No new leaf nodes found in the taxonomy.

### ISSUE THREAD

**Issue by djkemmet** [instructlab/instructlab#2338]
**Describe the bug**
A Valid Toxonomy is throwing errors instead of being used to train data.

**To Reproduce**
1. Clone taxonomy
2. add your taxonomy data to one of the top level toxonomy dirs. In my case that was linguistics/dialectics which contained a qna.yaml file and attribution.txt file 
4. validate your taxonomy file with `ilab taxonomy diff --taxonomy-path ./taxonomy` and receive "Taxonomy in ./taxonomy is valid :)"
5. Now attempt to generate training data with ` ilab data generate --taxonomy-path ./taxonomy` and get error: Generating dataset failed with the following error: Error: No new leaf nodes found in the taxonomy.

**Expected behavior**
I expect that with `ilab taxonomy diff --taxonomy-path ./taxonomy` reports "Taxonomy in ./taxonomy is valid :)" that that taxonomy can be used without issue to generate training data. 

**Screenshots**
![image](https://github.com/user-attachments/assets/a897cc30-eaa0-4bac-b5f6-8524c2505a93)

**Device Info (please complete the following information):**
 - Hardware Specs: Apple M3 Max, 36GB Ram
 - OS Version: MacOS 14.6.1 (23G93)
 - Python Version: Python 3.11.10
 - InstructLab Version: 
 - sys.version: 3.11.10 (main, Sep  7 2024, 01:03:31) [Clang 15.0.0 (clang-1500.3.9.4)]
sys.platform: darwin
os.name: posix
platform.release: 23.6.0
platform.machine: arm64
platform.node: DJs-MacBook-Pro
platform.python_version: 3.11.10
instructlab.version: 0.18.4
instructlab-dolomite.version: 0.1.1
instructlab-eval.version: 0.1.2
instructlab-quantize.version: 0.1.0
instructlab-schema.version: 0.3.1
instructlab-sdg.version: 0.2.7
instructlab-training.version: 0.4.2
torch.version: 2.3.1
torch.backends.cpu.capability: NO AVX
torch.version.cuda: None
torch.version.hip: None
torch.cuda.available: False
torch.backends.cuda.is_built: False
torch.backends.mps.is_built: True
torch.backends.mps.is_available: True
llama_cpp_python.version: 0.2.79
llama_cpp_python.supports_gpu_offload: True
(venv) dj@DJs-MacBook-Pro instructlab % 

**Additional context**

**Comment by cdoern** [comment]
the `ilab taxonomy diff` cmd should also print which files you modified/added which it is not. Did you commit the changes? I believe they need to be uncommited changes to taxonomy.

**Comment by jaideepr97** [comment]
also i think the file has to be named "qna.yaml" (in case it is not)

**Comment by romeosdibm** [comment]
@djkemmet  any updates how this was fixed ?

**Comment by hupiper** [comment]
I'm getting the same thing on two different operating systems. Tried it on my Mac and on my Fedora laptops. The diff command is not picking up the changes.

**Comment by hupiper** [comment]
NM - I was used to writing ansible, and I named the file .yml. Once I renamed it to .yaml, it seemed to work.

**Comment by stevef1uk** [comment]
I have the same error. I deleted an entry rather than added one e.g. swifties 

I am running on Linux (6.11.0-13-generic #14-Ubuntu SMP PREEMPT_DYNAMIC Sat Nov 30 23:51:51 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux)

ilab, version 0.22.2
Python 3.11.11
Package                           Version
--------------------------------- -----------------
absl-py                           2.1.0
accelerate                        1.0.1
aiofiles                          24.1.0
aiohappyeyeballs                  2.4.4
aiohttp                           3.11.11
aiohttp-cors                      0.7.0
aiosignal                         1.3.2
airportsdata                      20241001
annotated-types                   0.7.0
anyio                             4.8.0
appdirs                           1.4.4
astor                             0.8.1
attrs                             24.3.0
beautifulsoup4                    4.12.3
bitsandbytes                      0.45.0
blake3                            1.0.2
cachetools                        5.5.0
certifi                           2024.12.14
chardet                           5.2.0
charset-normalizer                3.4.1
click                             8.1.8
click-didyoumean                  0.3.1
cloudpickle                       3.1.1
colorama                          0.4.6
colorful                          0.5.6
compressed-tensors                0.8.1
DataProperty                      1.1.0
datasets                          2.21.0
deepsearch-glm                    0.26.2
deepspeed                         0.16.2
depyf                             0.18.0
dill                              0.3.8
diskcache                         5.6.3
distlib                           0.3.9
distro                            1.9.0
docker-pycreds                    0.4.0
docling                           2.8.3
docling-core                      2.14.0
docling-ibm-models                2.0.8
docling-parse                     2.1.2
docstring_parser                  0.16
docutils                          0.21.2
easyocr                           1.7.2
einops                            0.8.0
et_xmlfile                        2.0.0
evaluate                          0.4.3
fastapi                           0.115.6
filelock                          3.16.1
filetype                          1.2.0
flash-attn                        2.7.3
frozenlist                        1.5.0
fsspec                            2024.6.1
gguf                              0.10.0
gitdb                             4.0.12
GitPython                         3.1.44
google-api-core                   2.24.0
google-auth                       2.37.0
googleapis-common-protos          1.66.0
grpcio                            1.69.0
h11                               0.14.0
hf_transfer                       0.1.9
hjson                             3.1.0
httpcore                          1.0.7
httptools                         0.6.4
httpx                             0.28.1
huggingface-hub                   0.27.1
idna                              3.10
imageio                           2.36.1
importlib_metadata                8.5.0
iniconfig                         2.0.0
instructlab                       0.22.2
instructlab-dolomite              0.2.0
instructlab-eval                  0.4.2
instructlab-quantize              0.1.0
instructlab-schema                0.4.2
instructlab-sdg                   0.6.3
instructlab-training              0.6.1
interegular                       0.3.3
Jinja2                            3.1.5
jiter                             0.8.2
joblib                            1.4.2
jsonlines                         3.1.0
jsonpatch                         1.33
jsonpointer                       3.0.0
jsonref                           1.1.0
jsonschema                        4.23.0
jsonschema-specifications         2024.10.1
langchain-core                    0.3.29
langchain-text-splitters          0.3.5
langsmith                         0.2.10
lark                              1.2.2
lazy_loader                       0.4
linkify-it-py                     2.0.3
llama_cpp_python                  0.3.2
llvmlite                          0.43.0
lm_eval                           0.4.7
lm-format-enforcer                0.10.6
lxml                              5.3.0
markdown-it-py                    3.0.0
marko                             2.1.2
MarkupSafe                        3.0.2
mbstrdecoder                      1.1.3
mdit-py-plugins                   0.4.2
mdurl                             0.1.2
memray                            1.15.0
mistral_common                    1.5.1
more-itertools                    10.6.0
mpmath                            1.3.0
msgpack                           1.1.0
msgspec                           0.19.0
multidict                         6.1.0
multiprocess                      0.70.16
nest-asyncio                      1.6.0
networkx                          3.4.2
ninja                             1.11.1.3
nltk                              3.9.1
numba                             0.60.0
numexpr                           2.10.2
numpy                             1.26.4
nvidia-cublas-cu12                12.1.3.1
nvidia-cuda-cupti-cu12            12.1.105
nvidia-cuda-nvrtc-cu12            12.1.105
nvidia-cuda-runtime-cu12          12.1.105
nvidia-cudnn-cu12                 9.1.0.70
nvidia-cufft-cu12                 11.0.2.54
nvidia-curand-cu12                10.3.2.106
nvidia-cusolver-cu12              11.4.5.107
nvidia-cusparse-cu12              12.1.0.106
nvidia-ml-py                      12.560.30
nvidia-nccl-cu12                  2.20.5
nvidia-nvjitlink-cu12             12.4.127
nvidia-nvtx-cu12                  12.1.105
openai                            1.59.7
opencensus                        0.11.4
opencensus-context                0.1.3
opencv-python-headless            4.10.0.84
openpyxl                          3.1.5
orjson                            3.10.14
outlines                          0.0.46
outlines_core                     0.1.26
packaging                         24.2
pandas                            2.2.3
pandas-stubs                      2.2.3.241126
partial-json-parser               0.2.1.1.post5
pathspec                          0.12.1
pathvalidate                      3.2.3
peft                              0.9.0
pillow                            10.4.0
pip                               24.3.1
platformdirs                      4.3.6
pluggy                            1.5.0
portalocker                       3.1.1
prometheus_client                 0.21.1
prometheus-fastapi-instrumentator 7.0.2
prompt_toolkit                    3.0.48
propcache                         0.2.1
proto-plus                        1.25.0
protobuf                          4.25.5
psutil                            6.1.1
py-cpuinfo                        9.0.0
py-spy                            0.4.0
pyairports                        2.1.1
pyarrow                           18.1.0
pyasn1                            0.6.1
pyasn1_modules                    0.4.1
pybind11                          2.13.6
pyclipper                         1.3.0.post6
pycountry                         24.6.1
pydantic                          2.9.2
pydantic_core                     2.23.4
pydantic-settings                 2.7.1
pydantic_yaml                     1.4.0
Pygments                          2.19.1
pypdfium2                         4.30.1
pytablewriter                     1.2.1
pytest                            8.3.4
python-bidi                       0.6.3
python-dateutil                   2.9.0.post0
python-docx                       1.1.2
python-dotenv                     1.0.1
python-pptx                       1.0.2
pytz                              2024.2
PyYAML                            6.0.2
pyzmq                             26.2.0
ray                               2.40.0
referencing                       0.35.1
regex                             2024.11.6
requests                          2.32.3
requests-toolbelt                 1.0.0
rich                              13.9.4
rouge_score                       0.1.2
rpds-py                           0.22.3
rsa                               4.9
Rtree                             1.3.0
ruamel.yaml                       0.18.10
ruamel.yaml.clib                  0.2.12
sacrebleu                         2.5.1
safetensors                       0.5.2
scikit-image                      0.25.0
scikit-learn                      1.6.1
scipy                             1.15.1
sentencepiece                     0.2.0
sentry-sdk                        2.20.0
setproctitle                      1.3.4
setuptools                        75.8.0
shapely                           2.0.6
shellingham                       1.5.4
shortuuid                         1.0.13
shtab                             1.7.1
six                               1.17.0
smart-open                        7.1.0
smmap                             5.0.2
sniffio                           1.3.1
soupsieve                         2.6
sqlitedict                        2.1.0
sse-starlette                     2.2.1
starlette                         0.41.3
starlette-context                 0.3.6
sympy                             1.13.1
tabledata                         1.3.4
tabulate                          0.9.0
tcolorpy                          0.1.7
tenacity                          9.0.0
tesserocr                         2.7.1
textual                           1.0.0
threadpoolctl                     3.5.0
tifffile                          2025.1.10
tiktoken                          0.7.0
tokenizers                        0.21.0
toml                              0.10.2
torch                             2.4.0+cu121
torchaudio                        2.4.0+cu121
torchvision                       0.19.0+cu121
tqdm                              4.67.1
tqdm-multiprocess                 0.0.11
transformers                      4.48.0
triton                            3.0.0
trl                               0.13.0
typeguard                         4.4.1
typepy                            1.3.4
typer                             0.12.5
types-pytz                        2024.2.0.20241221
typing_extensions                 4.12.2
tyro                              0.9.9
tzdata                            2024.2
uc-micro-py                       1.0.3
urllib3                           2.3.0
uvicorn                           0.34.0
uvloop                            0.21.0
virtualenv                        20.28.1
vllm                              0.6.2
wandb                             0.16.6
watchfiles                        1.0.4
wcwidth                           0.2.13
websockets                        14.1
wheel                             0.45.1
word2number                       1.1
wrapt                             1.17.2
xdg-base-dirs                     6.0.2
xformers                          0.0.27.post2
xgrammar                          0.1.9
XlsxWriter                        3.2.0
xxhash                            3.5.0
yamllint                          1.35.1
yarl                              1.18.3
zipp                              3.21.0
zstandard                         0.23.0


(vinstruct) stevef@omen:~/dev/instructlab$ ilab taxonomy diff
Taxonomy in /home/stevef/dev/instructlab/taxonomy is valid :)
(vinstruct) stevef@omen:~/dev/instructlab$ ilab data generate --taxonomy-path /home/stevef/dev/instructlab/taxonomy 
INFO 2025-01-16 08:09:35,452 numexpr.utils:162: NumExpr defaulting to 16 threads.
INFO 2025-01-16 08:09:35,624 datasets:59: PyTorch version 2.4.0+cu121 available.
INFO 2025-01-16 08:09:36,503 instructlab.model.backends.llama_cpp:125: Trying to connect to model server at http://127.0.0.1:8000/v1
llama_new_context_with_model: n_ctx_per_seq (4096) < n_ctx_train (32768) -- the full capacity of the model will not be utilized
WARNING 2025-01-16 08:09:39,258 instructlab.data.generate_data:77: Disabling SDG batching - unsupported with llama.cpp serving
INFO 2025-01-16 08:09:39,262 instructlab.data.generate_data:87: Generating synthetic data using 'full' pipeline, '/home/stevef/.cache/instructlab/models/mistral-7b-instruct-v0.2.Q4_K_M.gguf' model, '/home/stevef/dev/instructlab/taxonomy' taxonomy, against http://127.0.0.1:59465/v1 server
failed to generate data with exception: Generating dataset failed with the following error: Error: No new leaf nodes found in the taxonomy.

**Comment by jwm4** [comment]
@stevef1uk , I think if you only deleted a node and didn't add or modify anything, then this is working as intended: SDG by default runs over only the elements of the taxonomy that were added or modified.   If you want SDG to run on everything in the taxonomy regardless of whether it was added or modified or there in the original that you checked out, then you specify `--taxonomy-base=empty` in the command line for `ilab data generate`.   That tells it to generate data for the whole taxonomy (as if you were starting from an *empty* taxonomy instead of just making an incremental addition to an existing one).

**Comment by hupiper** [comment]
Just coming back to this issue - if you let the yaml files have both .yml and .yaml extensions, my problem is solved.
